### PR TITLE
Provides a test to validate the end to end connectivity of routing::Route

### PR DESCRIPTION
**Pull request by agalbachicar**
# 🎉 New feature

Part of maliput#543
Pairs with https://github.com/maliput/maliput/pull/628

## Summary

Provides test coverage to https://github.com/maliput/maliput/pull/628 with the ability to incorporate a concrete maliput backend.

## Test it

Via unit tests

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if it affects the public API)

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by francocipollone — APPROVED** [review]
LGTM!

File: test/route_end_to_end_connectivity_test.cc [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+std::string SerializeErrors(const std::vector<std::string>& errors) {
+  std::string result;
+  for (const std::string& error : errors) {
+    result += error + " | ";
+  }
+  return result;
+}
+
+class MultipleStraightSegmentsTest : public ::testing::Test {
+ public:
```
Comment (liangfok) [inline test/route_end_to_end_connectivity_test.cc:75]: To be consistent with other tests in this directory, mark this `protected`.

File: test/route_end_to_end_connectivity_test.cc [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+  const api::LaneId kLane_1_0_1{"1_0_1"};
+  const api::LaneId kLane_1_0_m1{"1_0_-1"};
+  const api::LaneId kLane_2_0_1{"2_0_1"};
+  const api::LaneId kLane_2_0_m1{"2_0_-1"};
+  const api::Lane* lane_1_0_1;
+  const api::Lane* lane_1_0_m1;
+  const api::Lane* lane_2_0_1;
+  const api::Lane* lane_2_0_m1;
+
+  std::unique_ptr<api::RoadNetwork> road_network_{};
```
Comment (liangfok) [inline test/route_end_to_end_connectivity_test.cc:93]: Move this to the bottom of this section.
https://google.github.io/styleguide/cppguide.html#Declaration_Order

File: test/route_end_to_end_connectivity_test.cc [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+  static constexpr double kLaneLength{100.};
+  const api::LaneId kLane_1_0_1{"1_0_1"};
+  const api::LaneId kLane_1_0_m1{"1_0_-1"};
+  const api::LaneId kLane_2_0_1{"2_0_1"};
+  const api::LaneId kLane_2_0_m1{"2_0_-1"};
+  const api::Lane* lane_1_0_1;
+  const api::Lane* lane_1_0_m1;
+  const api::Lane* lane_2_0_1;
+  const api::Lane* lane_2_0_m1;
```
Comment (liangfok) [inline test/route_end_to_end_connectivity_test.cc:91]: The above four should end in "_{}" and be moved to the bottom of this section.

```
  const api::Lane* lane_1_0_1_{};
  const api::Lane* lane_1_0_m1_{};
  const api::Lane* lane_2_0_1_{};
  const api::Lane* lane_2_0_m1_{};
```

File: test/route_end_to_end_connectivity_test.cc [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+  const routing::Phase phase(0 /*index*/, kLinearTolerance, {start_position}, {end_position},
+                             {lane_s_range_m1, lane_s_range_1}, road_network_.get());
+  const routing::Route dut({phase}, road_network_.get());
+
+  const std::vector<std::string> errors = dut.ValidateEndToEndConnectivity();
+
+  ASSERT_TRUE(errors.empty()) << SerializeErrors(errors);
+}
+
+TEST_F(SinglePhaseTest, SinglePhaseWithTwoLanesAndStartPositionsIsNotConnectedEndToEnd) {
```
Comment (liangfok) [inline test/route_end_to_end_connectivity_test.cc:146]: Add a comment explaining why having start positions results in the route not being connected end-to-end.

File: test/route_end_to_end_connectivity_test.cc [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+  const routing::Phase phase(0 /*index*/, kLinearTolerance, {start_position_1, start_position_m1}, {end_position},
+                             {lane_s_range_m1, lane_s_range_1}, road_network_.get());
+  const routing::Route dut({phase}, road_network_.get());
+
+  const std::vector<std::string> errors = dut.ValidateEndToEndConnectivity();
+
+  ASSERT_FALSE(errors.empty()) << SerializeErrors(errors);
+}
+
+TEST_F(SinglePhaseTest, SinglePhaseWithTwoLanesAndEndPositionsIsNotConnectedEndToEnd) {
```
Comment (liangfok) [inline test/route_end_to_end_connectivity_test.cc:161]: Ditto. Add a comment explaining why having end positions results in the route not being connected end-to-end.

File: test/route_end_to_end_connectivity_test.cc [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+  const routing::Phase last_phase(1 /*index*/, kLinearTolerance, {last_start_position}, {last_end_position},
+                                  {last_lane_s_range}, road_network_.get());
+  const routing::Route dut({first_phase, last_phase}, road_network_.get());
+
+  const std::vector<std::string> errors = dut.ValidateEndToEndConnectivity();
+
+  ASSERT_TRUE(errors.empty()) << SerializeErrors(errors);
+}
+
+TEST_F(MultiPhaseTest, MultiPhaseWithOneLaneEachIsNotGeometricallyConnectedEndToEnd) {
```
Comment (liangfok) [inline test/route_end_to_end_connectivity_test.cc:196]: Ditto. Add a comment explaining why having phases with one lane each results in the route not being connected end-to-end.

(Ditto similar situations below.)

**Review by liangfok — CHANGES_REQUESTED** [review]
Completed a pass.

File: test/route_end_to_end_connectivity_test.cc (hunk 6 above)
Comment (agalbachicar) [reply test/route_end_to_end_connectivity_test.cc:199]: [5415ba7](https://github.com/maliput/maliput_integration_tests/pull/81/commits/5415ba7b4a80d037e1a4fe16999f64b6bfd528aa)

File: test/route_end_to_end_connectivity_test.cc (hunk 5 above)
Comment (agalbachicar) [reply test/route_end_to_end_connectivity_test.cc:163]: [5415ba7](https://github.com/maliput/maliput_integration_tests/pull/81/commits/5415ba7b4a80d037e1a4fe16999f64b6bfd528aa)

File: test/route_end_to_end_connectivity_test.cc (hunk 4 above)
Comment (agalbachicar) [reply test/route_end_to_end_connectivity_test.cc:147]: [5415ba7](https://github.com/maliput/maliput_integration_tests/pull/81/commits/5415ba7b4a80d037e1a4fe16999f64b6bfd528aa)

File: test/route_end_to_end_connectivity_test.cc (hunk 3 above)
Comment (agalbachicar) [reply test/route_end_to_end_connectivity_test.cc]: [5415ba7](https://github.com/maliput/maliput_integration_tests/pull/81/commits/5415ba7b4a80d037e1a4fe16999f64b6bfd528aa)

File: test/route_end_to_end_connectivity_test.cc (hunk 2 above)
Comment (agalbachicar) [reply test/route_end_to_end_connectivity_test.cc]: [5415ba7](https://github.com/maliput/maliput_integration_tests/pull/81/commits/5415ba7b4a80d037e1a4fe16999f64b6bfd528aa)

File: test/route_end_to_end_connectivity_test.cc (hunk 1 above)
Comment (agalbachicar) [reply test/route_end_to_end_connectivity_test.cc]: [5415ba7](https://github.com/maliput/maliput_integration_tests/pull/81/commits/5415ba7b4a80d037e1a4fe16999f64b6bfd528aa)

**Review by francocipollone — DISMISSED** [review]
LGTM!

File: test/route_end_to_end_connectivity_test.cc [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+#include <maliput_malidrive/builder/params.h>
+#include <maliput_malidrive/builder/road_network_builder.h>
+#include <maliput_malidrive/loader/loader.h>
+
+#include "assert_compare.h"
+
+// @file This file evaluates the end to end connectivity constraints of a maliput::routing::Route.
+// @details This is not done in maliput because of the need of a concrete maliput::api::RoadGeometry
+// to evaluate all the topological and geometrical constraints. Mocking that for the purposes of
+// this test would require implementing a concrete backend. In this test, an ad-hoc xodr map is used.
```
Comment (liangfok) [inline test/route_end_to_end_connectivity_test.cc:55]: > xodr

"*.xodr (OpenDRIVE)"

or

"OpenDRIVE"

**Review by liangfok — CHANGES_REQUESTED** [review]
Completed a pass.

File: test/route_end_to_end_connectivity_test.cc (hunk 7 above)
Comment (agalbachicar) [reply test/route_end_to_end_connectivity_test.cc]: [4832c43](https://github.com/maliput/maliput_integration_tests/pull/81/commits/4832c437a21a3942c02ad67e744ee793d407731e)

**Review by liangfok — APPROVED** [review]
LGTM

**Pull request merged**